### PR TITLE
Implement #5547: FileUpload: support chunking and resume

### PR DESCRIPTION
**Pull request by christophs78**
jQuery File Upload already supports chunked file upload.

Current state of implementation / limitations / ...
- [x] It´s only supported for mode="advanced"
- [x] ~~It´s only supported for native - Impl (NativeFileUploadDecoder). It should be possible to add this to CommonsFileUploadDecoder too.
Update: Implemented for CommonsFileUploadDecoder too.~~
- [x] ~~FileUpload-listener is called once per chunk. It´s up to the application to handle these chunks.~~
- [x] ~~NativeUploadedFile was extended with a few attributes which indicate which chunk was uploaded ->WIP~~
- [x] ~~VirusScanner-integration (after the whole file was uploaded?) -> not supported~~
- [x] ~~Whatˋs about ContentTypeVaildation via Tika? Does it need the whole file? Is it enough to look at the first (and/or last) chunk? -> not supported~~

TODOs
- [x] ~~i´ll provide a showcase-PR in a few days~~
- [x] CSP test with CSP enabled to make sure nothing is broken
- [x] automatic resume (which interval? how much retries? --> configurable)
- [x] (from chunks) merged file does not get removed automatically from tempdir
- [x] listener/chunkListener which one is called at which point in time?
- [x] remove chunk files (of aborted uploads) --> automatically at abort
- [x] add unique praefix to tempFile? (AbstractFileUploadDecoder#processLastChunk)

ShowCase-PR: https://github.com/primefaces/primefaces-showcase/pull/124

### CODE REVIEW

**Comment by Rapster** [comment]
I'm actually working on this in my company project. Will definitely keep an eye on your PR :wink:

File: src/main/java/org/primefaces/component/fileupload/NativeFileUploadDecoder.java
Comment (melloware) [inline src/main/java/org/primefaces/component/fileupload/NativeFileUploadDecoder.java]: Compile this pattern in a static final CONSTANT for performance so it doesn't have to recompile this pattern on every call.

File: src/main/java/org/primefaces/component/fileupload/NativeFileUploadDecoder.java
Comment (melloware) [inline src/main/java/org/primefaces/component/fileupload/NativeFileUploadDecoder.java]: Probably safe to throw an exception here as its an expected case.

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by Rapster** [comment]
> FileUpload-listener is called once per chunk. It´s up to the application to handle these chunks.

I think current `listener` should be called once file is fully uploaded. Now I wonder if we should also have a listener for chunks, first use case that comes on top of my head is digest calculation

File: src/main/java/org/primefaces/component/fileupload/NativeFileUploadDecoder.java
Comment (Rapster) [reply src/main/java/org/primefaces/component/fileupload/NativeFileUploadDecoder.java]: Based on https://tools.ietf.org/html/rfc7233#section-4.2, I would go for `^bytes (\d+)-(\d+)\/(\d+|\*)$`

**Comment by christophs78** [comment]
> I think current `listener` should be called once file is fully uploaded. Now I wonder if we should also have a listener for chunks, first use case that comes on top of my head is digest calculation

I´m not sure about this point yet....
1. It should be possible to provide a specialized listerner for chunks. (Thats the easy thing.)
2. What do we expect from the "classic" fully-uploaded listener in chunked-upload-mode? We try to avoid to keep the whole file in memory. Which information should the fully-uploaded listener provide in chunked-upload-mode?
3. What should VirusScan do in chunked-upload-mode? We OOTB never have the whole file. We only have chunks. IMO it does not make sense to only scan chunks. Should this be handled by the application? Or should PF chunked-file-upload write one condensed temporary file in a configureable file-system-location? Do VirusScan based on this and provide this file to the fully-uploaded listener?

**Comment by christophs78** [comment]
@Rapster: It´s still WIP but can you have a look at this PR? Any thoughts?

**Comment by melloware** [comment]
@christophs78 Looks like you have some minor conflicts.

File: docs/9_0/components/fileupload.md
Comment (melloware) [inline docs/9_0/components/fileupload.md:328]: Still need examples here?

File: src/main/java/org/primefaces/component/fileupload/FileUpload.java
Comment (melloware) [inline src/main/java/org/primefaces/component/fileupload/FileUpload.java]: You can remove the parenthesis around this.

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by Rapster** [comment]
Hey @christophs78,

Sorry for the late answer, but quite busy with work lately... Anyway, first and foremost: great job! I'm happy to see that feature implemented. Since I'm working on uploading large files as well, this will definitely help us to improve the way we upload files.

Though I have a couple of questions:
- Chunks are being merged by the end-user (based on your wiki doc) why that? I might be wrong but I would expect the component to do that by itself. 
- My idea of having a `chunkListener`  was to be able to do few things meantime the file is being uploaded (for example, computing digest file)
- I don't think chunks should be set as submittedValue, only the "final" file should be. Chunks are only temporary structures, they're not validated ,processed virus-scanned or anything so I wouldn't submit them as a proper value. Being said, that should simplify a lot your impl.
- Chunks upload should also be supported in "simple" mode, but since you're merging the final file through chunkListener, I understand why you can't.

One final thought, IMO the best thing about having chunk upload is to be able to resume an upload (see https://github.com/blueimp/jQuery-File-Upload/wiki/Chunked-file-uploads#resuming-file-uploads). We can talk about it once we're clear on how chunks should be handled by the component

**Comment by christophs78** [comment]
Hi @Rapster!

Thanks for your feedback!

IMO there are two motiviations for chuncked file upload:
- Reduce the serverside memory-footprint when uploading large files (eg 1GB)
- Provide resume functionality. (Primarily for slow or not stable internet connections.)

My efforts to this point concentrate on the serverside memory-footprint because this is a issue for the Ops-colleagues in my organisation. 
My chunked - implementation is somewhat basic. We "stream" the (large) uploaded file(s) to the server. And the implementation on the server has to decide how it likes to further process the chunks.

Maybe we need something like to modes for chunked file upload:
- One which concentrates one low serverside memory-footprint (which is implemented in this PR) at the expense of increased developer-complexity.
- One which concentrates on resume-functionality and keeps a simple API at the expense of increased memory-footprint.

Do you see a way to catch both requirements with one implementation?

**Comment by christophs78** [comment]
Is it an option to merge this PR after review and work on the resume-part in a separate PR?
(I already started playing around with resume.)

**Comment by tandraschko** [comment]
It would be great if you can first agree with @rapster about the feature set and resolve the open tasks :)

**Comment by Rapster** [comment]
Hi everyone,

Let me first answer to your last statement @christophs78:
> Reduce the serverside memory-footprint when uploading large files (eg 1GB)

Not sure what memory you're referring to, if it's RAM then chunks use more memory than a one long sequential upload (and that's okay! unless your chunks are 1GB 😅 ). Chunks are saved in memory (according to Jetty at least) while no-chunk write a file through InputStream. Though, network congestion is more likely to happen with no-chunk.

So basically, IMO what we want to solve with chunks are:
- avoiding network congestion
- and just like you said, the resume functionnality. It's actually the very first reason we want this, since it solves a lot of problem caused by long uploads (e.g network instability, timeouts etc.)

@christophs78 I can submit a PR on your branch to share my thoughts regarding my previous comments.

**Comment by christophs78** [comment]
@Rapster: You are right. I tried to solve a phantom-issue. We had issues with uploading large files (up to 1 GB), where our application-server took a multiple of the filesize as memory.
Then we read the wrong internet-sources and my colleagues and i went into to wrong direction. 
And we did not debug what realy happens inside the application server / JSF / PrimeFaces until now. 
(Our real issue must have been the processing of the uploaded file.)

Even in our environment with OpenLiberty the upload-handling seems to be fine OOTB:
image

So i´m totally happy when you submit a PR based on my branch.

**Comment by Rapster** [comment]
I'll try to push something this week, I managed to make chunks and resume working

File: docs/9_0/components/fileupload.md
Comment (melloware) [inline docs/9_0/components/fileupload.md]: Still a TODO?

**Review by melloware — CHANGES_REQUESTED** [review]

**Comment by christophs78** [comment]
Many thanks to @Rapster who did great things in his "nightshift".
Tried to keep Checkstyle happy and added a bit of JSDoc to keep this happy too. (Still some things to learn regarding TypeScript and JSDoc. At least for me.)
I´ll come back and update the showcase-PR later this day.

File: src/main/java/org/primefaces/component/fileupload/FileUpload.java
Comment (Rapster) [inline src/main/java/org/primefaces/component/fileupload/FileUpload.java]: chunks are never submitted as a value (and that's on purpose), so this condition will never be verified

**Comment by Rapster** [comment]
Thanks @christophs78 so far I think it's a good first draft, in a few words:
- chunk and resume features are supported
- chunks are not submitted as values
- merge chunks is supported by the component out of the box (but it's possible for the user to define is own through SPI)
- tika content type is not verified for chunks (but as mentionned @christophs78 earlier we could test the first chunk)

Though I'm not happy with one thing: `FileUploadChunkDecoder#getUploadDirectory`
We should be able to retrieve the upload directory from apache commons or servlet 3.0, and not hardcoded like I did. I'm still testing with servlet 3.0 to make sure I understand how the MultiPartConfig works, but so far it's quite confusing...

File: src/main/java/org/primefaces/component/fileupload/DefaultFileUploadChunkDecoder.java
Comment (Rapster) [inline src/main/java/org/primefaces/component/fileupload/DefaultFileUploadChunkDecoder.java]: The chunkDir is deleted right after the final file is merged, I'd rather delete the chunks in one shot after making sure the final file is merged

**Review by Rapster — CHANGES_REQUESTED** [review]

File: src/main/java/org/primefaces/component/fileupload/DefaultFileUploadChunkDecoder.java
Comment (christophs78) [reply src/main/java/org/primefaces/component/fileupload/DefaultFileUploadChunkDecoder.java]: Deleting chunkDir with (chunk-)files within throws a DirectoryNotEmptyException.

File: src/main/java/org/primefaces/component/fileupload/FileUpload.java
Comment (christophs78) [reply src/main/java/org/primefaces/component/fileupload/FileUpload.java:101]: IMO the core question is do we want to to have a chunkListener? Or are we fine with the „normal“ listener which get´s fired after the last chunk.
The code you wrote last night fired the „normal“ listener after each chunk. With my modification we fire chunkListener after each chunk and the „normal“ listener is fired after the last chunk was uploaded and we have the whole file on the server.
From my side i do not think we need the chunkListener. IMO ist kind of nice2have.

Comment (christophs78) [reply src/main/java/org/primefaces/component/fileupload/FileUpload.java]: IMO the core question is do we want to to have a chunkListener? Or are we fine with the „normal“ listener which get´s fired after the last chunk.
The code you wrote last night fired the „normal“ listener after each chunk. With my modification we fire chunkListener after each chunk and the „normal“ listener is fired after the last chunk was uploaded and we have the whole file on the server.
From my side i do not think we need the chunkListener. IMO ist kind of nice2have.

File: src/main/java/org/primefaces/component/fileupload/DefaultFileUploadChunkDecoder.java
Comment (Rapster) [reply src/main/java/org/primefaces/component/fileupload/DefaultFileUploadChunkDecoder.java]: Ah indeed... Last night I brought a lot of modifications just like this one, I guess it was too late to see it 😅 Anyway i'll have a look at it

File: src/main/java/org/primefaces/component/fileupload/FileUpload.java
Comment (Rapster) [reply src/main/java/org/primefaces/component/fileupload/FileUpload.java]: That's actually something have been wondering as well, do we need this chunkListener...? The only use case I have in mind is whenever you want and can calculate something over chunks (as mentionned digest for example) instead of processing the whole file. Have a look at https://docs.oracle.com/javase/7/docs/api/java/security/DigestInputStream.html to have a clearer of what I'm referring to.

Right now, that's probably something we don't really need, being said, the use case I mentionned might be solved using a custom `FileUploadChunkDecoder`

**Comment by christophs78** [comment]
@Rapster: Thanks for your work from late monday!

Just added auto-resume. What do you think about this? Should it be configureable via (eg) a chunkUploadConfig-object?

Can you have a look at our TODO-list at the top of this PR. Any thoughts on them? What should we implement? Should we discard some of them?

**Comment by Rapster** [comment]
I'll have to check the todo list. Have you ever had that issue: `Uncaught Error: cannot call methods on fileupload prior to initialization; attempted to call` calling `$.blueimp.fileupload.prototype.options.fail.call(this, e, data);`?

I see you're not using it while in the plugin example, it does. And I think we need it to support the auto upload feature

**Comment by christophs78** [comment]
> I'll have to check the todo list. Have you ever had that issue: `Uncaught Error: cannot call methods on fileupload prior to initialization; attempted to call` calling `$.blueimp.fileupload.prototype.options.fail.call(this, e, data);`?
> 
> I see you're not using it while in the plugin example, it does. And I think we need it to support the auto upload feature

Just added auto="true" to https://github.com/primefaces/primefaces-showcase/blob/a0b8155ea4c9d4f158b7a08ab05104dba13c0821/src/main/webapp/ui/file/upload/chunked.xhtml. Seems to work fine with FF 75 and Chrome 81.
You got this error once or is it reproduceable? All browsers?

**Comment by christophs78** [comment]
Added following things with my last commit:
- we create a subdirectory "PF_chunkedUpoad" below java.io.tmpdir and use this for upload (makes it easier for cleanup)
- the merged files get a UUID-prefix (users may upload files with the same names; these should not collide)
- a simple cleanup-mechanism which run´s once a day an removes all chunkDirs (from aborted uploads) which are older than one day; we may need to discuss this one, make it configureable, ...

**Comment by Rapster** [comment]
Sorry I wrote that in the morning but forgot to press the button 😅 :

Yeah just adding auto attribute seems to be working...

Regarding the todolist:

> - [ ] automatic resume (which interval? how much retries?)

Introduce 2 new attributes

> - [ ] (from chunks) merged file does not get removed automatically from tempdir

When do they get removed usually?

- [ ] remove chunk files (of aborted uploads) after eg 1 day

I'm working on it, but ideally the component shouldn't have to check ellapsed time of the file and then delete it, a cron job would be more appropriate (just like the author of the plugin mention)

- [ ] add UUID-praefix to tempFile to wholeFile? (AbstractFileUploadDecoder#processLastChunk)

Simply prefix with the X-File-Id such as `[X-File-Id]Filename.ext`.

**Comment by christophs78** [comment]
> > * (from chunks) merged file does not get removed automatically from tempdir
> 
> When do they get removed usually?

When i think about the temp files for upload handled by the servlet container (servlet 3.0), they are IMO removed after the request.
So i think about a PF-filter which implements javax.servlet.Filter, remembers the merged file in a static ThreadLocal-variable and removes it at the end of the request.

Plan B may be something like the developer has to remove the merged file manually after it´s processed.

**Comment by christophs78** [comment]
Tomcat 9.0.12 throws the following error during startup:
01-May-2020 10:41:58.677 SCHWERWIEGEND [RMI TCP Connection(5)-127.0.0.1] com.sun.faces.config.ConfigureListener.contextInitialized Critical error during deployment: 
 java.lang.UnsupportedOperationException: **Section 4.4 of the Servlet 3.0 specification does not permit this method to be called from a ServletContextListener that was not defined in web.xml, a** web-fragment.xml file nor annotated with @WebListener
	at org.apache.catalina.core.StandardContext$NoPluggabilityServletContext.addListener(StandardContext.java:6593)
	at **org.primefaces.webapp.PostConstructApplicationEventListener.addUploadedFileCleanerListener(PostConstructApplicationEventListener.java:65)**
	at org.primefaces.webapp.PostConstructApplicationEventListener.processEvent(PostConstructApplicationEventListener.java:60)
	at javax.faces.event.SystemEvent.processListener(SystemEvent.java:123)
	at com.sun.faces.application.applicationimpl.Events.processListeners(Events.java:253)
	at com.sun.faces.application.applicationimpl.Events.invokeListenersFor(Events.java:231)
	at com.sun.faces.application.applicationimpl.Events.publishEvent(Events.java:112)
	at com.sun.faces.application.ApplicationImpl.publishEvent(ApplicationImpl.java:127)
	at javax.faces.application.ApplicationWrapper.publishEvent(ApplicationWrapper.java:788)
	at com.sun.faces.config.ConfigManager.publishPostConfigEvent(ConfigManager.java:548)
	at com.sun.faces.config.ConfigureListener.contextInitialized(ConfigureListener.java:265)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4641)
...

File: src/main/resources/META-INF/resources/primefaces/fileupload/2-fileupload.js
Comment (christophs78) [inline src/main/resources/META-INF/resources/primefaces/fileupload/2-fileupload.js]: Not sure whether this should be at the start of fail-function. Think this would delete already uploaded chunks before we have the chance to resume.

Comment (Rapster) [reply src/main/resources/META-INF/resources/primefaces/fileupload/2-fileupload.js]: Not sure what you mean? fail function is called when user abort upload

**Comment by Rapster** [comment]
Thanks for reporting that issue on Tomcat, I guess I can't add it programmatically that way then, so it'll be up to the user to add that listener, which is not a bad thing by the way since I don't fully understand why uploaded files should be right away deleted

File: src/main/resources/META-INF/resources/primefaces/fileupload/2-fileupload.js
Comment (christophs78) [reply src/main/resources/META-INF/resources/primefaces/fileupload/2-fileupload.js]: Fail is also called when the network-connection breaks. (eg when i simulate this via browser-devtools) 
We may move this block into the `if (data.errorThrown === 'abort')` { a few lines down.

**Comment by christophs78** [comment]
> Thanks for reporting that issue on Tomcat, I guess I can't add it programmatically that way then, so it'll be up to the user to add that listener, which is not a bad thing by the way since I don't fully understand why uploaded files should be right away deleted

Is it a bad idea to add https://docs.oracle.com/javaee/7/api/javax/servlet/annotation/WebListener.html - annotation which exists since Servlet 3.0 or ship web-fragment.xml as part of PrimeFaces? (so only a few people still using 2.x - Servlet - containers would need to handle this themself)
Servlet 3.0 was introduced 2009. So i would hope most of our PF-userbase uses a servlet container which supports this. Or am I too optimistic?

File: src/main/resources/META-INF/web-fragment.xml
Comment (Rapster) [inline src/main/resources/META-INF/web-fragment.xml:1]: If I'm not mistaken, web-fragment is servlet 3.0

**Comment by tandraschko** [comment]
We should add a web-fragment and add a docu hint for serlvet 2.5.
I wouldn't use annotations for a filter as they were broken partly in older containers.

File: docs/9_0/components/fileupload.md
Comment (christophs78) [inline docs/9_0/components/fileupload.md]: @Rapster : Can you help out with your english-language-skills? I´m not totally happy with my text.

**Comment by Rapster** [comment]
I think we good for now. Feel free to review that PR

**Comment by christophs78** [comment]
`UploadedFileCleanerListener` does not delete the merged file anymore.
It´s maybe a regression due to https://github.com/primefaces/primefaces/pull/5607/commits/0fd106ca4b4b3837106481907cbd600412b41266.

**Comment by Rapster** [comment]
Fixed :wink:

**Comment by christophs78** [comment]
I´m happy with this PR.
@melloware, @tandraschko : do you want to have a look at this PR?

**Review by melloware — APPROVED** [review]
I have been following this PR the whole time and it looks great.  this is a nice improvement to PF.

**Comment by tandraschko** [comment]
I wonder if we shouldn't use the servlet temp dir for uploading? (https://docs.oracle.com/javaee/6/api/javax/servlet/ServletContext.html#TEMPDIR)

**Comment by Rapster** [comment]
I just had a try, `request.getServletContext().getAttribute(ServletContext.TEMPDIR)` returns a tmp folder contained in deployed app (target folder, e.g `D:\showcase-facelift\target\tmp`). I don't think we want to keep chunks in that location.

Besides that, I read this post https://stackoverflow.com/a/30111831/4605161 and BalusC answer. Made me realize that Servlet 3.0 is JEE 6.0, so how come do we still support Servlet 2.5? Being said, it does not mean we should remove code like CommonsFileUploader, but do we support Servlet 2.5 anywhere else?

**Comment by tandraschko** [comment]
Actually the servlet-tempdir feels IMO more correct as the java tempdir.

We currently do support Servlet2.5 as there is not much difference. Just some auto-register features like web-fragment are missing but therefore we added it in the docu.
I think we could add a servlet25 flag in the PrimeConfig and therefore use servlet-tempdir or java-tempdir.

**Comment by Rapster** [comment]
Yeah but see, even though there is not much of a difference between 2.5 and 3.0, PF codebase will still have to make a few adjustments to support 2.5 (so more tricks to detects which Servlet version are we running on...) Moreover, which does not really make clear where PF stands on which environnement it should run, JEE 6 but compatible Servlet 2.5? Sounds a bit confusing to me... I would rather keep things simple and clear, simply by saying it's JEE 6 therefore Servlet 3.0.

Regarding tempdir, I guess there are pros and cons, more cons than pros I think according to this: https://stackoverflow.com/a/18664715/4605161 But i'm opened to discuss what's best @melloware @christophs78

**Comment by tandraschko** [comment]
As long we can easily support it, we should do it (like with adding docu hint to add the listener in the web.xml instead of rely on the web-fragment). Same with JSF 2.0.
If this feature will never run on 2.5 because of missing API parts, it's ok for me aslong we add it as hint to the docu. 

We probably MUST do a "modern" version of PrimeFaces after JSF 3.0, to use new API parts then.
3.0 will change just change the packages, JSF 4.0 will be with new features/deprecations and so on.

Then we also can remove support for older EE-specs.
We will probably have 3 submodules of primefaces then:
- core (maybe only JS?)
- javax / legacay
- jakarta / modern

**Comment by christophs78** [comment]
> Regarding tempdir, I guess there are pros and cons, more cons than pros I think according to this: https://stackoverflow.com/a/18664715/4605161 But i'm opened to discuss what's best @melloware @christophs78

Just asked myself what my Ops-colleague would prefer. (He is a 100+ kilo man with a lot of temperament.) In our organisation we have a so called document-path (which is set via -D vm-argument) which points to a directory where applications may save their files.
So i would personally prefer to make tempdir configurable. And use something else (java.io.tmpdir) as fallback.
That would be similar to what https://primefaces.github.io/primefaces/8_0/#/components/fileupload?id=filter-configuration does.

Regarding supporting very old JSF-, Servlet-, ... versions. We have limited resources and IMO we should concentrate them on efforts that move PrimeFaces forward. (Running regression- and compatiblility-tests against Servlet 2.5 does not make much sense IMO. Servlet 3.0 was released in 2009)

As suggested by @tandraschko we should IMO do this cut for one the next PrimeFaces major releases.

**Comment by Rapster** [comment]
> In our organisation we have a so called document-path (which is set via -D vm-argument) which points to a directory where applications may save their files.
> So i would personally prefer to make tempdir configurable. And use something else (java.io.tmpdir) as fallback.
> That would be similar to what https://primefaces.github.io/primefaces/8_0/#/components/fileupload?id=filter-configuration does.

It's already existing, either overriding `java.io.tmpdir` (which is more less meant for that) in your VM argument or this way:
```xml
<servlet>
        <servlet-name>Faces Servlet</servlet-name>
        <servlet-class>javax.faces.webapp.FacesServlet</servlet-class>

        <multipart-config>
            <location>D:\Test</location>
        </multipart-config>
    </servlet>
```

I don't consider chunks as a  "proper" uploaded file, they're just temporary data, so located in tmpdir fits more that purpose

**Comment by tandraschko** [comment]
Im sure you can also set the servlet tempdir in Tomcat e.g.

**Comment by Rapster** [comment]
You could, but that would change one thing: your war would be deployed into that directory. I don't think this is what we want

**Comment by Rapster** [comment]
Mmmh I'll dig deeper regarding this constant, you're right that might be the right way to go 😉 i'll let you know

**Comment by tandraschko** [comment]
then lets to for a context-param and fallback to the java tempdir?

**Comment by Rapster** [comment]
Alright, I kept things very simple, chunks and merged file are located in same parent folder
1. Use `javax.servlet.context.tempdir` as primary location
2. If defined, it'll use either the dir from apache commons or servlet 3.0
3. If none of this is set, then fallback to `java.io.tmpdir` (which should probably never happen)

**Comment by tandraschko** [comment]
Thanks for your effort - looks really good me to.
I will merge it now. If you can further improve it, just send a PR :D

**Comment by Rapster** [comment]
Thank you Thomas, for such commit you should use the squash and merge option, I was planning to do the merge myself but you were faster than me :sweat_smile:

**Pull request merged**
